hathor/websocket/streamer.py

```diff
@@ -102,87 +135,130 @@ def set_ack(self, ack: int) -> None:
+            return
+        self._state = StreamerState.CLOSED
         self.protocol.unregisterProducer()
         self.deferred.callback(success)
 
+    def gracefully_close(self) -> None:
+        """Gracefully close the stream by sending the StreamEndMessage and waiting for its ack."""
+        if not self._state.can_transition_to(StreamerState.CLOSING):
+            return
+        self.send_message(StreamEndMessage(id=self.stream_id, seq=self.get_next_seq()))
+        self._state = StreamerState.CLOSING
+
     def pauseProducing(self) -> None:
         """Pause streaming. Called by twisted."""
-        self._paused = True
+        if not self._state.can_transition_to(StreamerState.PAUSED):
+            self.protocol.log.warn('pause requested in an unexpected state', state=self._state)
+            return
+        self._state = StreamerState.PAUSED
+        self._is_paused_by_transport = True
 
     def stopProducing(self) -> None:
         """Stop streaming. Called by twisted."""
-        self._stop = True
+        if not self._state.can_transition_to(StreamerState.CLOSED):
+            self.protocol.log.warn('stopped requested in an unexpected state', state=self._state)
+            return
         self.stop(False)
 
     def resumeProducing(self) -> None:
         """Resume streaming. Called by twisted."""
-        self._paused = False
-        self._run()
-
-    def _run(self) -> None:
-        """Run the streaming main loop."""
-        coro = self._async_run()
-        Deferred.fromCoroutine(coro)
+        if not self._state.can_transition_to(StreamerState.ACTIVE):
+            self.protocol.log.warn('resume requested in an unexpected state', state=self._state)
+            return
+        self._is_paused_by_transport = False
+        self.resume_if_possible()
 
     def should_pause_streaming(self) -> bool:
+        """Return true if the streaming should pause due to the flow control mechanism."""
         if self._sliding_window_size is None:
             return False
         stop_value = self._last_ack + self._sliding_window_size + 1
         if self._next_sequence_number < stop_value:
             return False
         return True
 
+    def _run(self) -> None:
+        """Run the streaming main loop."""
+        if not self._state.can_transition_to(StreamerState.ACTIVE):
+            self.protocol.log.warn('_run() called in an unexpected state', state=self._state)
+            return
+        coro = self._async_run()
+        Deferred.fromCoroutine(coro)
+
     async def _async_run(self):
-        assert not self._is_running
-        self._is_running = True
+        assert not self._is_main_loop_running
+        self._state = StreamerState.ACTIVE
+        self._is_main_loop_running = True
         try:
             await self._async_run_unsafe()
         finally:
-            self._is_running = False
+            self._is_main_loop_running = False
```

Comment:
Should this mean the state is not `ACTIVE` anymore?

Reply:
`ACTIVE` indicates that the streamer is actively attempting to send messages. However, the execution of the main loop is somewhat independent of this state. The main loop can stop due to flow control mechanisms, meaning that the streamer may be in the `ACTIVE` state without the main loop actually running.

Although I considered introducing separate states like `ACTIVE_NOT_RUNNING` and `ACTIVE_RUNNING` to differentiate these scenarios, I ultimately decided to maintain a single `ACTIVE` state, using a supporting flag to track whether the main loop is currently running.